### PR TITLE
Improve ModuleExpansionTransformer performance

### PR DESCRIPTION
**Pull request by jonjohnsonjr**
The existing implementation took a very long time given a large number of Vertices or Modules in a graph.

This constructs a tree that can quickly return the set of vertices in a given module rather than iterating over every Vertex for every Module.

We have a fairly unusual graph with an obscene number of modules and vertices, so this change makes a fairly substantial impact on our performance.

Before: https://share.firefox.dev/3Y0qu9h
After: https://share.firefox.dev/4bHYBGd

Time spent in ModuleExpansionTransformer drops from ~27s to ~1s.

I haven't added any new tests because (in theory) this should not affect any functionality, just performance, but let me know if I should add some.

This `pathTree` implementation is not exposed to anything because I didn't want to change any interfaces, but it might make sense for [`tofu.Graph`](https://github.com/opentofu/opentofu/blob/f668c48ffd57d1f5f5dcd7543f179b373643081a/internal/tofu/graph.go#L23) to implement something like this? I would personally wait until we find other instances where we want vertex lookup by module before changing any APIs, but I'd defer to maintainers.

## Target Release

1.9.0

## Checklist

- [x] I have read the [contribution guide](https://github.com/opentofu/opentofu/blob/main/CONTRIBUTING.md).
- [x] I have not used an AI coding assistant to create this PR.
- [x] I have written all code in this PR myself OR I have marked all code I have not written myself (including modified code, e.g. copied from other places and then modified) with a comment indicating where it came from.

### Go checklist

 

- [x] I have run golangci-lint on my change and receive no errors relevant to my code.
- [x] I have run existing tests to ensure my code doesn't break anything.
- [x] I have added tests for all relevant use cases of my code, and those tests are passing.
- [x] I have only exported functions, variables and structs that should be used from other packages.
- [x] I have added meaningful comments to all exported functions, variables, and structs.

### Website/documentation checklist

- [ ] I have locally started the website as [described here](https://github.com/opentofu/opentofu/blob/main/website/README.md) and checked my changes.

### CODE REVIEW

**Comment by cam72cam** [comment]
Hi @jonjohnsonjr, thanks for the PR!  There is a lot of low hanging fruit ripe to be optimized in the graph transformation.

> graph with an obscene number of modules and vertices

Are you able to build a test scenario with a provider like `tfcoremock`?  As we review more of these much needed optimization PRs, we are starting to build up a suite of performance tests to simulate some of the larger and more complex configurations in use.

**Comment by jonjohnsonjr** [comment]
> Are you able to build a test scenario with a provider like tfcoremock? As we review more of these much needed optimization PRs, we are starting to build up a suite of performance tests to simulate some of the larger and more complex configurations in use.

I'll try to put something together, it might take me a bit.

**Comment by jonjohnsonjr** [comment]
Please forgive my ignorance here, I couldn't figure out a way to easily generate addressable resources without resorting to bash, I assume there's a better way to do this, but:

My `main.tf`:
```
module "generated-1" {
  source = "./generated"
}

module "generated-2" {
  source = "./generated"
}

module "generated-3" {
  source = "./generated"
}

module "generated-4" {
  source = "./generated"
}

module "generated-5" {
  source = "./generated"
}

module "generated-6" {
  source = "./generated"
}

module "generated-7" {
  source = "./generated"
}

module "generated-8" {
  source = "./generated"
}

module "generated-9" {
  source = "./generated"
}

module "generated-10" {
  source = "./generated"
}

```

In `submodule/main.tf`:
```
resource "tfcoremock_simple_resource" "iter" {
  count  = 1
  string = count.index
}
```

I then generated 1000 modules in `./generated/main.tf` that reference `./submodule`:

```bash
mkdir generated
for i in $(seq 1 1000); do echo module \"module-$i\" \{ source = \"../submodule\" \}; done > ./generated/main.tf
```

Before:
```
tofu plan  91.57s user 10.24s system 268% cpu 37.926 total
```

After:
```
tofu plan  57.09s user 9.55s system 352% cpu 18.882 total
```

**Comment by cam72cam** [comment]
Hey, sorry for the late reply!  I'll dig into this tomorrow!  Test scenario looks good!

**Comment by jonjohnsonjr** [comment]
Opened https://github.com/opentofu/opentofu-performance-testing/pull/1

When validating this I also noticed there was another n^2 loop for connecting the closers just below, so I've fixed that as well.

File: internal/command/terraform.tfstate [hunk 1]
```diff
@@ -0,0 +1 @@
+{"version":4,"terraform_version":"1.9.0","serial":0,"lineage":"fake-for-testing","outputs":{"foo":{"value":"bar","type":"string","sensitive":true}},"resources":[],"check_results":null}
```
Comment (cam72cam) [inline internal/command/terraform.tfstate:1]: Looks like this was accidentally committed.

**Comment by cam72cam** [comment]
Looks like some crashes in the tests

File: internal/tofu/transform_module_expansion.go [hunk 2]
```diff
@@ -75,17 +85,26 @@ func (t *ModuleExpansionTransformer) Transform(g *Graph) error {
...
-			if len(d.Addr) > len(c.Addr) && c.Addr.Equal(d.Addr[:len(c.Addr)]) {
-				g.Connect(dag.BasicEdge(c, d))
+		// For a closer c with address ["module.foo", "module.bar", "module.baz"], we'll look up:
+		//
+		// t.closers["module.foo"]
+		// t.closers["module.foo.module.bar"]
+		//
+		// And connect the parent module to c.
+		//
+		// We skip i=0 because c.Addr[0:0].String() == "".
```
Comment (cam72cam) [inline internal/tofu/transform_module_expansion.go:95]: Perhaps we could clarify this note to mention that the root module ("") will not exist in t.closers and can be skipped?

File: internal/tofu/transform_module_expansion.go [hunk 3]
```diff
@@ -75,17 +85,26 @@ func (t *ModuleExpansionTransformer) Transform(g *Graph) error {
 	// Modules implicitly depend on their child modules, so connect closers to
 	// other which contain their path.
 	for _, c := range t.closers {
-		for _, d := range t.closers {
-			if len(d.Addr) > len(c.Addr) && c.Addr.Equal(d.Addr[:len(c.Addr)]) {
-				g.Connect(dag.BasicEdge(c, d))
+		// For a closer c with address ["module.foo", "module.bar", "module.baz"], we'll look up:
```
Comment (cam72cam) [inline internal/tofu/transform_module_expansion.go:88]: we'll look up all potential parent modules.

File: internal/command/terraform.tfstate (hunk 1 above)
Comment (jonjohnsonjr) [reply internal/command/terraform.tfstate]: Oh how embarrassing, thanks.

File: internal/tofu/transform_module_expansion.go (hunk 2 above)
Comment (jonjohnsonjr) [reply internal/tofu/transform_module_expansion.go]: Done.

**Comment by cam72cam** [comment]
Tests look good, I ran them against main for the same effect.

One minor change needed is to add the standard license headers to your test file.

**Comment by Yantrio** [comment]
I'm happy to approve this as long as the license headers are updated 👍

**Comment by jonjohnsonjr** [comment]
> One minor change needed is to add the standard license headers to your test file.

Whoops, copied them over from the test file I cribbed. Thanks!

**Review by cam72cam — APPROVED** [review]

**Comment by jonjohnsonjr** [comment]
Friendly ping @Yantrio

**Comment by cam72cam** [comment]
Thanks @jonjohnsonjr !

**Pull request merged**